scipy/special/specfun.c

```diff
@@ -0,0 +1,8418 @@
+
+    const double eps = 1.0e-15;
+    const double pi = 3.141592653589793;
+    const double c1 = 0.355028053887817;
+    const double c2 = 0.258819403792807;
+    const double sr3 = 1.732050807568877;
+
+    km2 = 0;
+    xa = fabs(x);
+    xq = sqrt(xa);
+    xm = 8.0;
+    if (x > 0.0) xm = 5.0;
+
+    if (x == 0.0) {
+        *ai = c1;
+        *bi = sr3 * c1;
+        *ad = -c2;
+        *bd = sr3 * c2;
+        return;
+    }
+
+    if (xa <= xm) {
+        fx = 1.0;
+        r = 1.0;
+        for (k = 1; k <= 40; k++) {
+            r *= x / (3.0 * k) * x / (3.0 * k - 1.0) * x;
+            fx += r;
+            if (fabs(r) < fabs(fx) * eps) break;
+        }
+
+        gx = x;
+        r = x;
+        for (k = 1; k <= 40; k++) {
+            r *= x / (3.0 * k) * x / (3.0 * k + 1.0) * x;
+            gx += r;
+            if (fabs(r) < fabs(gx) * eps) break;
+        }
+
+        *ai = c1 * fx - c2 * gx;
+        *bi = sr3 * (c1 * fx + c2 * gx);
+
+        df = 0.5 * x * x;
+        r = df;
+        for (k = 1; k <= 40; k++) {
+            r *= x / (3.0 * k) * x / (3.0 * k + 2.0) * x;
+            df += r;
+            if (fabs(r) < fabs(df) * eps) break;
+        }
+
+        dg = 1.0;
+        r = 1.0;
+        for (k = 1; k <= 40; k++) {
+            r *= x / (3.0 * k) * x / (3.0 * k - 2.0) * x;
+            dg += r;
+            if (fabs(r) < fabs(dg) * eps) break;
+        }
+
+        *ad = c1 * df - c2 * dg;
+        *bd = sr3 * (c1 * df + c2 * dg);
+    } else {
+        km = (int)(24.5 - xa);
+        if (xa < 6.0) km = 14;
+        if (xa > 15.0) km = 10;
+
+        if (x > 0.0) {
+            kmax = km;
+        } else {
+            // Choose cutoffs so that the remainder term in asymptotic
+            // expansion is epsilon size. The X<0 branch needs to be fast
+            // in order to make AIRYZO efficient
+            km2 = km;
+            if (xa > 70.0) km = 3;
+            if (xa > 500.0) km = 2;
+            if (xa > 1000.0) km = 1;
+
+            km2 = km;
+            if (xa > 150.0) km2 = 1;
+            if (xa > 3000.0) km2 = 0;
+            kmax = 2 * km + 1;
+        }
+        xe = xa * xq / 1.5;
+        xr1 = 1.0 / xe;
+        xar = 1.0 / xq;
+        xf = sqrt(xar);
+        rp = 0.5641895835477563;
+        r = 1.0;
+        for (k = 1; k <= kmax; k++) {
+            r *= (6.0 * k - 1.0) / 216.0 * (6.0 * k - 3.0)
+                / k * (6.0 * k - 5.0) / (2.0 * k - 1.0);
+            ck[k - 1] = r;
+            dk[k - 1] = -(6.0 * k + 1.0) / (6.0 * k - 1.0) * r;
+        }
+
+        if (x > 0.0) {
+            sai = 1.0;
+            sad = 1.0;
+            sbi = 1.0;
+            sbd = 1.0;
+            r = 1.0;
+            for (k = 1; k <= km; k++) {
+                r *= -xr1;
+                sai += ck[k - 1] * r;
+                sad += dk[k - 1] * r;
+                sbi += ck[k - 1] * r;
+                sbd += dk[k - 1] * r;
+            }
+            xp1 = exp(-xe);
```

Comment:
Maybe the two loops shouldn't be merged.
Because R has a different sign.

https://github.com/scipy/scipy/blob/c2806e4ec7651fbfa401829884476a9036988579/scipy/special/specfun/specfun.f#L7041-L7053

Reply:
Even now, I had to look two times to see the minus sign, I'll fix.